Userland/Libraries/LibIMAP/Parser.cpp

```diff
@@ -0,0 +1,836 @@
+/*
+ * Copyright (c) 2021, Kyle Pereira <hey@xylepereira.me>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Parser.h"
+#include <AK/CharacterTypes.h>
+
+namespace IMAP {
+
+ParseStatus Parser::parse(ByteBuffer&& buffer, bool expecting_tag)
+{
+    if (m_incomplete) {
+        m_buffer += buffer;
+        m_incomplete = false;
+    } else {
+        m_buffer = move(buffer);
+        position = 0;
+        m_response = SolidResponse();
+    }
+
+    if (match("+")) {
+        consume(" ");
+        auto data = parse_while(+[](u8 x) { return x != '\r'; });
+        consume("\r\n");
+        return { true, { ContinueRequest { data } } };
+    }
+
+    while (match("*")) {
+        parse_untagged();
+    }
+
+    if (expecting_tag) {
+        if (end()) {
+            m_incomplete = true;
+            return { true, {} };
+        }
+        parse_response_done();
+    }
+
+    if (m_parsing_failed) {
+        return { false, {} };
+    } else {
+        return { true, { { move(m_response) } } };
+    }
+}
+
+bool Parser::match(StringView x)
+{
+    size_t i = 0;
+    auto previous_position = position;
+    while (i < x.length() && !end() && to_ascii_lowercase(x[i]) == to_ascii_lowercase(m_buffer[position])) {
+        i++;
+        position++;
+    }
+    if (i != x.length()) {
+        // we didn't match the full string.
+        position = previous_position;
+        return false;
+    }
+
+    return true;
+}
+
+void Parser::parse_response_done()
+{
+    consume("A");
+    auto tag = parse_number();
+    consume(" ");
+
+    ResponseStatus status = parse_status();
+    consume(" ");
+
+    m_response.m_tag = tag;
+    m_response.m_status = status;
+
+    StringBuilder response_data;
+
+    while (!end() && m_buffer[position] != '\r') {
+        response_data.append((char)m_buffer[position]);
+        position += 1;
+    }
+
+    consume("\r\n");
+    m_response.m_response_text = response_data.build();
+}
+
+void Parser::consume(StringView x)
+{
+    if (!match(x)) {
+        dbgln("{} not matched at {}, buffer: {}", x, position, StringView(m_buffer.data(), m_buffer.size()));
+
+        m_parsing_failed = true;
+    }
+}
+
+// returns -1 if nothing matched.
+unsigned Parser::parse_number()
```

Comment:
`Optional<unsigned>`?
Also, you don't need to manually parse, just make a StringView and call `to_uint()` on it.